### DISCUSSION TITLE
CVS Validation - eForms Notice Editor Demo

### DISCUSSION THREAD

**Discussion by AlexAS81** [OP-TED/eForms-SDK#341] [Q&A] [answered]
We are trying to use the CVS validation functionality included in the eForms Notice Editor Demo application but it gives us the following error:

`{"timestamp":"2023-02-15T14:26:04.111+00:00","status":500,"error":"Internal Server Error","trace":"eu.europa.ted.eforms.noticeeditor.helper.validation.CvsApiException: Message=CVS POST response error: reason=, status=403, statusCode=403\r\n\tat eu.europa.ted.eforms.noticeeditor.helper.validation.CvsApiClient$1.handleResponse(CvsApiClient.java:219)\r\n\tat eu.europa.ted.eforms.noticeeditor.helper.validation.CvsApiClient$1.handleResponse(CvsApiClient.java:183)\r\n\tat org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:223)\r\n\tat org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:165)\r\n\tat org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:140)\r\n\tat eu.europa.ted.eforms.noticeeditor.helper.validation.CvsApiClient.httpPostToCvs(CvsApiClient.java:229)\r\n\tat eu.europa.ted.eforms.noticeeditor.helper.validation.CvsApiClient.validateNoticeXml(CvsApiClient.java:147)\r\n\tat eu.europa.ted.eforms.noticeeditor.service.NoticeValidationService.validateNoticeXmlUsingCvs(NoticeValidationService.java:117)\r\n\tat eu.europa.ted.eforms.noticeeditor.service.XmlWriteService.validateUsingCvs(XmlWriteService.java:165)\r\n\tat eu.europa.ted.eforms.noticeeditor.controller.XmlRestController.saveNoticeAndCvsValidate(XmlRestController.java:59)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\r\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\r\n\tat org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)\r\n\tat org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150)\r\n\tat org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117)\r\n\tat org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)\r\n\tat org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)\r\n\tat org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)\r\n\tat org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1067)\r\n\tat org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963)\r\n\tat org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)\r\n\tat org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)\r\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:517)\r\n\tat org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)\r\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:584)\r\n\tat org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)\r\n\tat org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1631)\r\n\tat org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:292)\r\n\tat org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)\r\n\tat org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)\r\n\tat org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)\r\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)\r\n\tat org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)\r\n\tat org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)\r\n\tat org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)\r\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)\r\n\tat org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)\r\n\tat org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)\r\n\tat org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:96)\r\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)\r\n\tat org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)\r\n\tat org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)\r\n\tat org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)\r\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)\r\n\tat org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)\r\n\tat org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)\r\n\tat org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:548)\r\n\tat org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)\r\n\tat org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)\r\n\tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)\r\n\tat org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)\r\n\tat org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)\r\n\tat org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)\r\n\tat org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)\r\n\tat org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)\r\n\tat org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)\r\n\tat org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)\r\n\tat org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)\r\n\tat org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)\r\n\tat org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)\r\n\tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)\r\n\tat org.eclipse.jetty.server.Server.handle(Server.java:516)\r\n\tat org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)\r\n\tat org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)\r\n\tat org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)\r\n\tat org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)\r\n\tat org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)\r\n\tat org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)\r\n\tat org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)\r\n\tat org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)\r\n\tat org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)\r\n\tat java.base/java.lang.Thread.run(Thread.java:834)\r\n","message":"eu.europa.ted.eforms.noticeeditor.helper.validation.CvsApiException: Message=CVS POST response error: reason=, status=403, statusCode=403","path":"/xml/notice/save/validation/cvs"}`

What is the correct way to test this functionality?

**Comment by rouschr** [comment, marked as answer]
You have this message: `CVS POST response error: reason=, status=403`
You need to create a key for CVS usage, see:

https://docs.ted.europa.eu/api/index.html
https://docs.ted.europa.eu/api/endpoints/cvs-ted-europa-eu.html

See editor demo `application.yaml` for guidance:
```
client:
  cvs:
    url: "https://cvs.ted.europa.eu"
    ### Security: use a command line parameter for security related data
    ### Example -Dclient.cvs.apiKey=ENVIRONMENT_VARIABLE_VALUE_HERE
    apiKey: "YOUR_API_KEY"
```
When running the editor demo from command line you have to pass the parameter:
` ... jar ... -Dclient.cvs.apiKey=ENVIRONMENT_VARIABLE_VALUE_HERE`

Something like this:
`java -jar \
 -Dclient.cvs.apiKey=${CVS_API_KEY} \
 -Dproxy.url=${PROXY_URL} \  <--- if you are behind a proxy
 target/notice-editor-demo-XYZ.jar
`
or you can directly put the key after the `=`, but for security reasons it is best you place it there using an environment variable to avoid leaving traces in your command line history.

Another thing to know is that until SDK 1.6.0 (soon) is released the validation will fail as the XML tags will have a wrong sort order.
You can still setup your key and try it out to be prepared.

The code to fix this is already in place but it requires a special SDK `notice-types/notice-types.json` file to work.
You can preview it here: [notice-types.zip](https://github.com/OP-TED/eForms-SDK/files/10755180/notice-types.zip)
You can try to replace your local notice-types.json of SDK 1.5 by this file, but you would also have to trick the editor into thinking it is an SDK 1.6.x... it is easier to wait for SDK 1.6.0, the XML tag sorting feature is not supported for SDK 1.5.x.
There is more information about the namespaces which allows the code to find the correct xsd files, all of that is done by the Java code on the back end of the editor demo. Sorting of XML tags is performed at the end right before the creation of the XML text.
It is possible that this XML sorting code will be moved into the core project later for reuse by more projects.
At least you will be well ahead in your understanding of how it works, which is the point of putting these latest developments in the `develop` branch.

**Reply by AlexAS81** [reply]
Thanks, we'll try it.

**Reply by AlexAS81** [reply]
I have generated my API Key at https://developer.preview.ted.europa.eu/home and I am starting the application as follows:

java -jar -Dclient.cvs.apiKey=MY_API_KEY C:\eforms-editor\eforms-notice-editor\target\notice-editor-demo-0.0.1-SNAPSHOT.jar

And when trying to validate a file in the eForms Notice Editor application, it gives me the following error:

{"timestamp":"2023-02-16T14:54:25.395+00:00","status":500,"error":"Internal Server Error","path":"/xml/notice/save/validation /cvs"}

**Reply by AlexAS81** [reply]
When I try to test https://cvs.ted.europa.eu/swagger-ui/index.html#/notice-rest-controller/validate by adding my API-Key I get the following error when running the validation:

{
  "errorCode": 403,
  "message": "Invalid key in header: X-API-Key",
  "timestamp": "2023-02-16T15:16:08Z"
}

![image](https://user-images.githubusercontent.com/115476802/219408729-870d557f-181e-4e31-93d7-ad231b9ab9f8.png)

**Reply by rouschr** [reply]
> When I try to test https://cvs.ted.europa.eu/swagger-ui/index.html#/notice-rest-controller/validate by adding my API-Key I get the following error when running the validation: 403

I do not think I can help you further as this is outside of the scope of the editor demo and these github discussions if it also does not work directly through the API. You could try to see with the **helpdesk**: 
https://simap.ted.europa.eu/contact

**Reply by AlexAS81** [reply]
We have created a new API-Key and today it works fine (with the swagger and the EForms Editor).
We have tested both with correct sample XMLs from GitHub, because the XMLs generated in the EForms Editor are not properly sorted.

**Reply by bertrand-lorentz** [reply]
> I have generated my API Key at https://developer.preview.ted.europa.eu/home and

> When I try to test https://cvs.ted.europa.eu/swagger-ui/index.html#/notice-rest-controller/validate by adding my API-Key


You have created your API in the "Preview" environement, but you are using CVS in the "Production" environment. API keys are only valid for the environment they were created in, see: https://docs.ted.europa.eu/api/index.html#_testing_your_apps
Your key should work with CVS at https://cvs.preview.ted.europa.eu/swagger-ui/

**Reply by rouschr** [reply]
> because the XMLs generated in the EForms Editor are not properly sorted.

The editor demo sort feature is only executed when using 1.6.x which will be released soon.
